File: examples/canonical-terminal/go-cobra-bubbletea/internal/core/stats.go
package core

import "math"

func ComputeStats(jobs []Job) Stats {
	byStatus := map[string]int{
		string(StatusPending): 0,
		string(StatusRunning): 0,
		string(StatusDone):    0,
		string(StatusFailed):  0,
	}
	var durations []float64
	for _, job := range jobs {
		byStatus[string(job.Status)]++
		if job.DurationS != nil {
			durations = append(durations, *job.DurationS)
		}
	}
	avg := 0.0
	if len(durations) > 0 {
		sum := 0.0
		for _, value := range durations {
			sum += value
		}
		avg = math.Round(sum/float64(len(durations))*100) / 100
	}
	failureRate := 0.0
	if len(jobs) > 0 {
		failureRate = math.Round(float64(byStatus[string(StatusFailed)])/float64(len(jobs))*100) / 100
	}
	return Stats{
		Total:        len(jobs),
		ByStatus:     byStatus,
		AvgDurationS: avg,
		FailureRate:  failureRate,
	}
}

File: examples/canonical-terminal/typescript-yargs-blessed/src/core/stats.ts
import type { Job, JobStatus, Stats } from "./models.js";

const STATUSES: JobStatus[] = ["pending", "running", "done", "failed"];

export function computeStats(jobs: Job[]): Stats {
  const byStatus = STATUSES.reduce(
    (accumulator, status) => ({ ...accumulator, [status]: 0 }),
    {} as Record<JobStatus, number>,
  );
  const durations: number[] = [];

  for (const job of jobs) {
    byStatus[job.status] += 1;
    if (typeof job.duration_s === "number") {
      durations.push(job.duration_s);
    }
  }

  const avgDuration =
    durations.length > 0
      ? Number((durations.reduce((total, value) => total + value, 0) / durations.length).toFixed(2))
      : null;
  const failureRate = jobs.length > 0 ? Number((byStatus.failed / jobs.length).toFixed(2)) : 0;

  return {
    total: jobs.length,
    by_status: byStatus,
    avg_duration_s: avgDuration,
    failure_rate: failureRate,
  };
}
